pygeo/DVGeometryVSP.py

```diff
@@ -610,378 +782,427 @@ def writePlot3D(self, fileName):
+        are small, but if we end up doing the FD with results from another processor
+        the difference is large enough to completely mess up the sensitivity. Due to this
+        we compute both baseline point and perturbed point on the processor that perturbs a given DV
+        this is slightly slower but avoids this issue. the final gradient has some error still,
+        but much more managable and unimportant compared to errors introduced by FD itself.
+        See issue https://github.com/mdolab/pygeo/issues/58 for updates.
+        """
+
+        # timing stuff:
+        t1 = time.time()
+        tvsp = 0
+        teval = 0
+        tcomm = 0
+
+        # counts
+        nDV = self.getNDV()
+        dvKeys = list(self.DVs.keys())
+        nproc = self.comm.size
+        rank = self.comm.rank
+
+        # arrays to collect local pointset info
+        ul = numpy.zeros(0)
+        vl = numpy.zeros(0)
+        gl = numpy.zeros(0, dtype="intc")
+
+        for ptSetName in self.pointSets:
+            # initialize the Jacobians
+            self.pointSets[ptSetName].jac = numpy.zeros((3 * self.pointSets[ptSetName].nPts, nDV))
+
+            # first, we need to vstack all the point set info we have
+            # counts of these are also important, saved in ptSet.nPts
+            ul = numpy.concatenate((ul, self.pointSets[ptSetName].u))
+            vl = numpy.concatenate((vl, self.pointSets[ptSetName].v))
+            gl = numpy.concatenate((gl, self.pointSets[ptSetName].geom))
+
+        # now figure out which proc has how many points.
+        sizes = numpy.array(self.comm.allgather(len(ul)), dtype="intc")
+        # displacements for allgather
+        disp = numpy.array([numpy.sum(sizes[:i]) for i in range(nproc)], dtype="intc")
+        # global number of points
+        nptsg = numpy.sum(sizes)
+        # create a local new point array. We will use this to get the new
+        # coordinates as we perturb DVs. We just need one (instead of nDV times the size)
+        # because we get the new points, calculate the jacobian and save it right after
+        ptsNewL = numpy.zeros(len(ul) * 3)
+
+        # create the arrays to receive the global info
+        ug = numpy.zeros(nptsg)
+        vg = numpy.zeros(nptsg)
+        gg = numpy.zeros(nptsg, dtype="intc")
+
+        # Now we do an allGatherv to get a long list of all pointset information
+        self.comm.Allgatherv([ul, len(ul)], [ug, sizes, disp, MPI.DOUBLE])
+        self.comm.Allgatherv([vl, len(vl)], [vg, sizes, disp, MPI.DOUBLE])
+        self.comm.Allgatherv([gl, len(gl)], [gg, sizes, disp, MPI.INT])
+
+        # we now have all the point info on all procs.
+        tcomm += time.time() - t1
+
+        # We need to evaluate all the points on respective procs for FD computations
+
+        # allocate memory
+        pts0 = numpy.zeros((nptsg, 3))
+
+        # evaluate the points
+        for j in range(nptsg):
```

Comment:
Yeah I think so. Because we just use rigid links between the actual points and the projected points, we can just evaluate the sensitivities at the projected points. That is why those pts0 is the projected points, evaluated with whatever is in vsp currently. The pointset update is completely separate.

I believe I used the convention: pts for projected points, and "points" for the actual tracked coordinates. This Jacobian works on the pts. Does that make sense? Feel free to resolve this conversation if you are happy with my explanation, we can dig further if you want.